scheduler/scheduler-examples/src/main/java/org/ow2/proactive/scheduler/examples/FTPConnector.java

```diff
@@ -0,0 +1,374 @@
+
+        getOut().println("LISTING directory: " + localParentDir);
+
+        File localDir = new File(localParentDir);
+        File[] subFiles = localDir.listFiles();
+        if (subFiles != null && subFiles.length > 0) {
+            for (File item : subFiles) {
+                String remoteFilePath = Paths.get(remoteDirPath, remoteParentDir, item.getName()).toString();
+                if (remoteParentDir.equals("")) {
+                    remoteFilePath = Paths.get(remoteDirPath, item.getName()).toString();
+                }
+
+                if (item.isFile()) {
+                    // upload the file
+                    String localFilePath = item.getAbsolutePath();
+                    getOut().println("About to upload the file: " + localFilePath);
+                    boolean uploaded = uploadSingleFile(ftpClient, localFilePath, remoteFilePath);
+                    if (uploaded) {
+                        getOut().println("UPLOADED a file to: " + remoteFilePath);
+                    } else {
+                        getOut().println("COULD NOT upload the file: " + localFilePath);
+                    }
+                } else {
+                    // create directory on the server
+                    boolean created = ftpClient.makeDirectory(remoteFilePath);
+                    if (created) {
+                        getOut().println("CREATED the directory: " + remoteFilePath);
+                    } else {
+                        getOut().println("COULD NOT create the directory: " + remoteFilePath);
+                    }
+
+                    // upload the sub directory
+                    String parent = Paths.get(remoteParentDir, item.getName()).toString();
+                    if (remoteParentDir.equals("")) {
+                        parent = item.getName();
+                    }
+
+                    localParentDir = item.getAbsolutePath();
+                    uploadDirectory(ftpClient, remoteDirPath, localParentDir, parent);
+                }
+            }
+        }
+    }
+
+    private static boolean uploadSingleFile(FTPClient ftpClient, String localFilePath, String remoteFilePath)
+            throws IOException {
+        File localFile = new File(localFilePath);
+
+        InputStream inputStream = new FileInputStream(localFile);
+        try {
+            ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
+            return ftpClient.storeFile(remoteFilePath, inputStream);
+        } finally {
+            inputStream.close();
+        }
+    }
+
+    private boolean checkDirectoryExists(FTPClient ftpClient, String dirPath) throws IOException {
+        ftpClient.changeWorkingDirectory(dirPath);
+        int returnCode = ftpClient.getReplyCode();
+        if (returnCode == 550) {
+            return false;
+        }
+        return true;
+    }
+
+    private void downloadDirectory(FTPClient ftpClient, String parentDir, String currentDir, String saveDir)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 29 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)